ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -216,6 +241,158 @@ KEXT_STATIC void UseMainForkIfNamedStream(
     }
 }
 
+KEXT_STATIC bool InitPendingRenames()
+{
+    if (version_major >= 18) // Only need to track renames on Mojave and newer
+    {
+        s_renameLock = SpinLock_Alloc();
+        s_maxPendingRenames = 8; // Arbitrary choice, should be maximum number of expected concurrent threads performing renames, but array will resize on demand
+        s_pendingRenameCount = 0;
+        s_pendingRenames = Memory_AllocArray<PendingRenameOperation>(s_maxPendingRenames);
+        if (!SpinLock_IsValid(s_renameLock) || s_pendingRenames == nullptr)
+        {
+            return false;
+        }
+
+        Array_DefaultInit(s_pendingRenames, s_maxPendingRenames);
+    }
+    
+    return true;
+}
+
+KEXT_STATIC void CleanupPendingRenames()
+{
+    if (version_major >= 18)
+    {
+        if (SpinLock_IsValid(s_renameLock))
+        {
+            SpinLock_FreeMemory(&s_renameLock);
+        }
+
+        if (s_pendingRenames != nullptr)
+        {
+            assert(s_pendingRenameCount == 0);
+            Memory_FreeArray(s_pendingRenames, s_maxPendingRenames);
+            s_pendingRenames = nullptr;
+            s_maxPendingRenames = 0;
+        }
+    }
+}
+
+KEXT_STATIC void ResizePendingRenames(uint32_t newMaxPendingRenames)
+{
+    PendingRenameOperation* newArray = Memory_AllocArray<PendingRenameOperation>(newMaxPendingRenames);
+    assert(newArray != nullptr);
+    PendingRenameOperation* arrayToFree = nullptr;
+    uint32_t arrayToFreeLength = 0;
+    
+    SpinLock_Acquire(s_renameLock);
+    {
+        if (newMaxPendingRenames > s_maxPendingRenames)
+        {
+            Array_CopyElements(newArray, s_pendingRenames, s_maxPendingRenames);
+            Array_DefaultInit(newArray + s_maxPendingRenames, newMaxPendingRenames - s_maxPendingRenames);
+            
+            arrayToFree = s_pendingRenames;
+            arrayToFreeLength = s_maxPendingRenames;
+            
+            s_pendingRenames = newArray;
+            s_maxPendingRenames = newMaxPendingRenames;
+        }
+        else
+        {
+            arrayToFree = newArray;
+            arrayToFreeLength = newMaxPendingRenames;
+        }
+    }
+    SpinLock_Release(s_renameLock);
+    
+    if (arrayToFree != nullptr)
+    {
+        assert(arrayToFreeLength > 0);
+        Memory_FreeArray(arrayToFree, arrayToFreeLength);
+    }
+}
+
+static void RecordPendingRenameOperation(vnode_t vnode)
+{
+    assertf(version_major >= 18, "KAUTH_FILEOP_WILL_RENAME is only supported by Darwin 18 and newer (version_major = %u)", version_major);
```

Comment:
It looks like we added some backwards compat in this PR, but due to this line I think we panic if we aren't on 18+.

Reply:
There’s a version check before calling this function. The first iteration of this PR was also backwards compatible but had a second commit which removed that compatibility. That broke the functional tests however, and I’ve not pushed it on the rebased version for now.

Reply:
Since this only gets called from` KAUTH_FILEOP_WILL_RENAME` you should be on 18+ if you get here.
Maybe an extra comment: `// This function should be called from KAUTH_FILEOP_WILL_RENAME`

Reply:
I assumed that was implicit in the assert message, but I can make that a bit clearer.

Reply:
Thanks @pmj.  I think a comment saying
// This function i called from `KAUTH_FILEOP_WILL_RENAME` a 18+ feature would be helpful

Reply:
I've updated the assert message to be more descriptive, no point splitting the information between the assert and a comment.